webapi/src/main/scala/org/knora/webapi/messages/admin/responder/projectsmessages/ProjectsMessagesADM.scala

```diff
@@ -505,130 +511,104 @@ case class ProjectADM(
-    )
+  final case class IriIdentifier(value: ProjectIri) extends ProjectIdentifierADM
+  object IriIdentifier {
+    def fromString(value: String): Validation[ValidationException, IriIdentifier] =
+      ProjectIri.make(value).map {
+        IriIdentifier(_)
+      }
+  }
 
   /**
-   * Returns an optional value of the identifier.
+   * Represents [[ShortcodeIdentifier]] identifier.
+   *
+   * @param value that constructs the identifier in the type of [[ShortCode]] value object.
    */
-  def toShortnameOption: Option[String] =
-    maybeShortname
+  final case class ShortcodeIdentifier(value: ShortCode) extends ProjectIdentifierADM
+  object ShortcodeIdentifier {
+    def fromString(value: String): Validation[ValidationException, ShortcodeIdentifier] =
+      ShortCode.make(value).map {
+        ShortcodeIdentifier(_)
+      }
+  }
 
   /**
-   * Tries to return the value as an SHORTCODE.
+   * Represents [[ShortnameIdentifier]] identifier.
+   *
+   * @param value that constructs the identifier in the type of [[ShortName]] value object.
    */
-  def toShortcode: String =
-    maybeShortcode.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.SHORTCODE' type.")
-    )
+  final case class ShortnameIdentifier(value: ShortName) extends ProjectIdentifierADM
+  object ShortnameIdentifier {
+    def fromString(value: String): Validation[ValidationException, ShortnameIdentifier] =
+      ShortName.make(value).map {
+        ShortnameIdentifier(_)
+      }
+  }
 
   /**
-   * Returns an optional value of the identifier.
+   * Represents [[UuidIdentifier]] identifier.
+   *
+   * @param value that constructs the identifier in the type of [[Base64Uuid]] value object.
    */
-  def toShortcodeOption: Option[String] =
-    maybeShortcode
+  final case class UuidIdentifier(value: Base64Uuid) extends ProjectIdentifierADM
+  // TODO-mpro: UuidIdentifier isn't fully used yet since the value object needs
+  object UuidIdentifier {
+    def fromString(value: String): Validation[ValidationException, UuidIdentifier] =
+      Base64Uuid.make(value).map {
+        UuidIdentifier(_)
+      }
+
+    def makeProjectIri(uuid: String) = s"http://rdfh.ch/projects/${uuid}"
+  }
 
   /**
-   * Returns the string representation
+   * Gets used identifier value.
+   *
+   * @param identifier either IRI, Shortname, Shortcode or UUID of the project.
+   * @return identifier's value as [[String]]
    */
-  override def toString: IRI =
-    s"ProjectIdentifierADM(${this.value})"
-
-}
-
-/**
- * Project identifier types:
- *  - IRI
- *  - Shortcode
- *  - Shortname
- */
-sealed trait ProjectIdentifierType
-object ProjectIdentifierType {
-  case object IRI       extends ProjectIdentifierType
-  case object SHORTCODE extends ProjectIdentifierType
-  case object SHORTNAME extends ProjectIdentifierType
+  def getId(identifier: ProjectIdentifierADM): String =
```

Comment:
To get an ID from an Identifier seems like non-sense to me. Can this be called differently, something like `getProjectIdType` and call the `ProjectIdentifierADM` `ProjectIdType` instead? I think this would make things easier to understand for others / in the future.

Reply:
Or maybe `getIdAsString`? Do whatever you think is best, you know best what the context is in which this method is used.